### ISSUE TITLE
Any interest in help migrating to TypeScript?

### ISSUE THREAD

**Issue by StephenFluin** [tobspr/shapez.io#522]
Love the game. Bought it on steam, would have paid more now that I know it's open source.

Are you interested in moving to TypeScript?

**Comment by tobspr** [comment]
Thanks! You can always get it on itch.io for a price you choose yourself :)

It's been a tough decision Between TS and Js but for now I am pretty happy with the combination of JS + jsdoc + tslint. It gives me all advantages without having to transpile it. But feel free to change my mind! :)

**Comment by StephenFluin** [comment]
Using what you know is definitely best. It looks like you are already using JSDoc typings though, which are much less readable, and much less strict, leading to more bugs (at least I make bugs like this). I have found that the transpilation almost never gets in the way once you have it setup.

I just compare the following:

```
    /**
     * Returns if this acceptor can accept a new item at slot N
     * @param {number} slotIndex
     * @param {BaseItem=} item
     */
    canAcceptItem(slotIndex, item) {
        const slot = this.slots[slotIndex];
        return this.filterMatches(slot.filter, item);
    }
```

and the equivalent TypeScript

```
    /**
     * Returns if this acceptor can accept a new item at slot N
     */
    canAcceptItem(slotIndex: number, item: BaseItem) {
        const slot = this.slots[slotIndex];
        return this.filterMatches(slot.filter, item);
    }
```

**Comment by dengr1065** [comment]
I'd really like to see that. TypeScript has lots of amazing features other than being strict-typed language, and code would be way more clean.

**Comment by tobspr** [comment]
Hm, well I think now its too late - maybe for the next project! :)

**Comment by Yona-Appletree** [comment]
As someone who has migrated large projects to TypeScript, it's never too late. One of the of major benefits of TypeScript is how it can be done incrementally, and really, it's just not that hard to add typings assuming your code is fairly clean (which this is). I totally agree that TypeScript would make this project easier to read and maintain -- but I'm a TS dev by day and am no doubt biased. Great work either way.

**Comment by akupiec** [comment]
I unfortunately agree with @Yona-Appletree 
Similar to him Im working on daily basic using TS (it's been already around 2 years), but I use it only in some my projects when I see benefits exactly due to transpilation time (I'm build time freek 100ms is to long when it's unnecessary ^_^ )

Im not not telling that TS it gold solution for all world problems but in some cases it's good to have types.

Moreover I have to say I wasn't always fan of Typescipt especially in the early versions when Draft was still alive :) but in current state it's totally worth migration.
When I migrated similar project (small library done in objective way) it instantly brought to light few typing bug's, and what's most important allowed me to vastly cleanup code due to better IDE integration and code jumping. 

**TL;DR:**
Migrate! You will get:
- rid of JSDoc and 98% of comments in the sources
- semi static types, coupled with interfaces, enums and decorators
- better code auto-complete / refactoring tools due to IDE integration
- increased building time, to get some coffee :D
